File: 514/b.c
#include <stdio.h>
typedef long long ll;

int main(){
  ll n,m;
 
  scanf("%lld",&n);
  getchar();
  scanf("%lld",&m);
  getchar();

  ll arr[n][m];
  ll newarr[n][m];

  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      newarr[i][j] = 0;
    }
  }
  int firstindex,secondindex,canput; 

  for(ll i=0;i<n;i++){
    for(ll j=0;j<m;j++){
      arr[i][j] = (getchar() == '#');
    }
    getchar();
  }

  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if (i==0 || j==0 || i==(n-1) || j==(m-1)) newarr[i][j] = 1;
      
      if(arr[i][j]==0){
        for(int q=-1; q<2;q++){
          for(int w=-1;w<2;w++){
            firstindex = i+q;
            secondindex = j+w;
            
            if ((q==0 && w==0)|| firstindex<0 || firstindex>(n-1) || secondindex<0 || secondindex>(m-1)) continue;
            newarr[firstindex][secondindex] = 1;
          }
        }
      }
    }
  }

/*
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      printf("%lld ",arr[i][j]);
    }
    putchar('\n');
  }

  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      printf("%lld ",newarr[i][j]);
    }
    putchar('\n');
  }
*/
  
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if(arr[i][j]==1){
        canput = 0;
        for(int q=-1; q<2;q++){
          for(int w=-1;w<2;w++){
            firstindex = i+q;
            secondindex = j+w;
            
            if ((q==0 && w==0)|| firstindex<0 || firstindex>(n-1) || secondindex<0 || secondindex>(m-1)) continue;

            if (newarr[firstindex][secondindex]==0){
              canput = 1;
              break;
            }
          }
          if (canput) break;
        }

        if (!canput){
          printf("NO\n");
          return 0;
        }
      }
    }
  }

  printf("YES\n");
}
        

File: 514/b.rb
nm = gets.chomp.split(" ").map(&:to_i)
n = nm[0]
m = nm[1]

arr = []

n.times do
  arr.push gets.chomp.split("").map{|z| z=="#"}
end

def nicep arr
  arr.each do |x|
    p x
  end
end

newarr = []

n.times do
  newarr.push [0]*m
end

(0..n-1).each do |i|
  (0..m-1).each do |j|
    if i==0 or j==0 or i==n-1 or j==m-1
      newarr[i][j] = 1
    end

    if arr[i][j] == false
      (0..2).each do |q|
        (0..2).each do |w|
          firstindex = i+q-1
          secondindex = j+w-1
          next if (q==1 and w==1) or (firstindex<0 or firstindex>(n-1) or secondindex<0 or secondindex>(m-1))
          newarr[firstindex][secondindex] = 1 #cannot be a center
        end
      end
    end
  end
end

#nicep newarr


(0..n-1).each do |i|
  (0..m-1).each do |j|
    
    if arr[i][j] == true
      canput = false

      (0..2).each do |q|
        (0..2).each do |w|
          firstindex = i+q-1
          secondindex = j+w-1
          next if (q==1 and w==1) or (firstindex<0 or firstindex>(n-1) or secondindex<0 or secondindex>(m-1))

=begin
          if (firstindex==1 and secondindex==1)
            puts "HI"
          end
=end
          if newarr[firstindex][secondindex] == 0 
            canput = true
            break
          end
        end
        break if canput
      end
      
#      puts "#{canput} #{i} #{j}"

      if not canput
        puts "NO"
        exit
      end
    end
  end
end

puts "YES"
